project/02-Modularity/Handout.md

```diff
@@ -0,0 +1,160 @@
+
+> :information_source: **This part of the project is rather long**. Next part will be way shorter, so you can spend about 1.5 weeks on this part.
+
+## Register to OpenWeatherMap
+
+[OpenWeatherMap](https://api.openweathermap.org/) offers an API that you can access over HTTPS. As this API offers both free and paid plans, they need to be able to identify the API caller, i.e. the App that is making the HTTPS requests. This is done using a secret key, that you provide in your requests to the service, and that identifies you uniquely.
+
+To get your secret key, you will need to register to [OpenWeatherMap](https://api.openweathermap.org/). Please have a look at their [Get Started](https://openweathermap.org/appid) page to do that.
+
+## Write down the structure of the system
+
+Before diving into the code, we ask you to first write down what the system you are designing will look like. 
+
+Please take a moment to think about the modules and interfaces you will need and what kind of methods they should expose to their client (the rest of your app). We advise you to write that down on a sheet of paper, or try to explain it to a rubber duck on your desk, to convince you that it works and will fulfill its purpose.
+
+Remember, what you will code will need to be able to retrieve the location of the user, retrieve the weather at a given location, as well as convert an address to a location.
+
+## Design and write the interfaces
+
+> :warning: We advise you not to read this section before thinking about the previous one, as we will "spoil" one of the possible structure of this system.
+
+Now that you know how your system should work, you can write interfaces and data classes that _describe_ this behaviour. Don't think too much about the actual implementation, just write an interface that exposes the behaviour you need.
+
+In theory, you will have three interfaces (the names are indicative):
+ - `LocationService` that defines a method to get the current position of the device
+ - `GeoCodingService` that defines a method to convert a position to an address (the reverse can also be useful if you want to display it to the user)
+ - `WeatherService` that defines a method to get the weather at a given location.
+
+You will also need a few classes to store the resulting forecast from the server. Think about _what values your client will need_ (while still making sure you can provide them - have a look at the [OneCall API documentation](https://openweathermap.org/api/one-call-api)). A `Location` class will also be needed to store the location - both when retrieving it in and when sending it to the weather API. Android already provides one, but you don't want to use it here, as it would not cleanly allow you to separate concerns and use an other provider in the future with the same interfaces.
```

Comment:
Think about _what values your client will need_ (while still making sure you can provide them - have a look at the [OneCall API documentation](https://openweathermap.org/api/one-call-api)).
->
Think about _which values your client will need_ (while still making sure the API can provide them - have a look at the [OneCall API documentation](https://openweathermap.org/api/one-call-api)).